tidy3d/plugins/smatrix/smatrix.py

```diff
@@ -0,0 +1,267 @@
+
+    def _shift_port(self, port: Port) -> Port:
+        """Generate a new port shifted by one grid cell in normal direction."""
+
+        shift_value = self._shift_value(port)
+        center_shifted = list(port.center)
+        center_shifted[port.size.index(0.0)] += shift_value
+        port_shifted = port.copy(deep=True)
+        port_shifted.center = center_shifted
+        return port_shifted
+
+    def _task_name(self, port_source: Port, mode_index: int) -> str:
+        """The name of a task, determined by the port of the source and mode index."""
+        return f"smatrix_port{port_source.name}_mode{mode_index}"
+
+    def _make_sims(self) -> Dict[str, Simulation]:
+        """Generate all the :class:`Simulation` objects for the S matrix calculation."""
+        sim_dict = {}
+        for port_source in self.ports:
+            for mode_source in self._to_sources(port_source):
+                sim_copy = self.simulation.copy(deep=True)
+                sim_copy.sources = [mode_source]
+                for port_monitor in self.ports:
+                    if port_source == port_monitor:
+                        port_monitor = self._shift_port(port_source)
+                    mode_monitor = self._to_monitor(port_monitor)
+                    sim_copy.monitors.append(mode_monitor)
+                    task_name = self._task_name(port_source, mode_source.mode_index)
+                    sim_dict[task_name] = sim_copy
+        return sim_dict
+
+    def _run_sims(
+        self, sim_dict: Dict[str, Simulation], folder_name: str, path_dir: str
+    ) -> "Batch":
+        """Run :class:`Simulations` for each port and return the batch after saving."""
+
+        # do it here as to not trigger web auth when importing the plugin
+        from ...web.container import Batch
+
+        batch = Batch(simulations=sim_dict, folder_name=folder_name)
+
+        # save to self, for reference later
+        self.batch = batch
+
+        batch.upload()
+
+        # save after upload so that the jobs are saved too for later
+        batch.to_file(self.batch_path)
+
+        batch.start()
+        batch.monitor()
+        return batch
+
+    def _normalization_factor(self, port_source: Port, sim_data: SimulationData) -> complex:
+        """Compute the normalization amplitude based on the measured input mode amplitude."""
+
+        port_monitor_data = sim_data[port_source.name]
+        mode_index = sim_data.simulation.sources[0].mode_index
+
+        normalize_amp = port_monitor_data.amps.sel(
+            f=self.freq,
+            direction=port_source.direction,
+            mode_index=mode_index,
+        ).values
+
+        normalize_n_eff = port_monitor_data.n_eff.sel(f=self.freq, mode_index=mode_index).values
+
+        k0 = 2 * np.pi * C_0 / self.freq
+        k_eff = k0 * normalize_n_eff
+        shift_value = self._shift_value(port_source)
+        return normalize_amp * np.exp(1j * k_eff * shift_value)
```

Comment:
the S_ii monitor is placed one grid cell away from the source to avoid numerical issues.  So this exp() stuff is to account for the phase difference between the source and monitor to compensate for this distance in a way that replicates what we would expect if they were at the exact same location.  Does that clear it up?

Reply:
I see, thanks